src/storage/txn/txn_status_cache.rs

```diff
@@ -301,93 +380,185 @@ impl TxnStatusCache {
+    pub fn upsert(&self, start_ts: TimeStamp, state: TxnState, now: SystemTime) {
         if !self.is_enabled {
             return;
         }
 
-        let insert_time = now.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
-        let mut slot = self.slots[self.slot_index(start_ts)].lock();
-        let previous_size = slot.size();
-        let previous_allocated = slot.internal_allocated_capacity();
-        slot.insert_if_not_exist(
-            start_ts,
-            CacheEntry {
-                commit_ts,
-                insert_time,
-            },
-        );
-        let size = slot.size();
-        let allocated = slot.internal_allocated_capacity();
-        // Update statistics.
-        // CAUTION: Assuming that only one TxnStatusCache instance is in a TiKV process.
+        let update_time = now.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
+        let entry = CacheEntry {
+            state: state.clone(),
+            update_time,
+        };
+        let slot_index = self.slot_index(start_ts);
+
+        let mut previous_size = 0;
+        let mut after_size = 0;
+        let mut previous_allocated = 0;
+        let mut after_allocated = 0;
+        match &state {
+            TxnState::Ongoing(min_commit_ts) => {
+                // The large_txn_cache takes precedence over the normal_cache in the retrieval
+                // process.
+                if *min_commit_ts > start_ts {
+                    // This case implies that min_commit_ts would be useful, thus the entry is put
+                    // in the large cache.
+                    let mut large_txn_cache = self.large_txn_cache[slot_index].lock();
+                    previous_size = large_txn_cache.size();
+                    previous_allocated = large_txn_cache.internal_allocated_capacity();
+                    if let Some(existing_entry) = large_txn_cache.get_mut(&start_ts) {
+                        // don't update committed or rolled back txns.
+                        if let TxnState::Ongoing(_) = existing_entry.state {
+                            existing_entry.state = state;
+                            existing_entry.update_time = update_time;
+                        }
+                    } else {
+                        large_txn_cache.insert(start_ts, entry);
+                    }
+                    after_size = large_txn_cache.size();
+                    after_allocated = large_txn_cache.internal_allocated_capacity();
+                } else {
+                    let mut normal_cache = self.normal_cache[slot_index].lock();
+                    previous_size = normal_cache.size();
+                    previous_allocated = normal_cache.internal_allocated_capacity();
+                    normal_cache.insert(start_ts, entry);
+                    after_size = normal_cache.size();
+                    after_allocated = normal_cache.internal_allocated_capacity();
+                }
```

Comment:
Is this branch the same as the `TxnState::Committed(_) | TxnState::RolledBack` branch? If so, consider put `if *min_commit_ts > start_ts` to the match arm `TxnState::Ongoing(min_commit_ts)`, so that the else branch and Committed/RolledBack branch can be merged together.

Reply:
good idea